Simulation customization userguide (rotating frames, RWA, sparse)

DanPuzzuoli opened the pull request:
### Summary

Closes #57 
Depends on #51 

Introduces a new userguide entry centered around "customizing" simulations via model transformations and evaluation modes. Phrases everything in terms of "how-to", e.g. "how-to use rotating frames, and their benefits", and demonstrates in a simple example the impact of making these choices. Also shows how to use sparse evaluation mode within JAX and emphasizes that when doing this, it is recommended to use diagonal frames only.

brosand commented on docs/userguide/how_to_configure_simulations.rst:
```diff
@@ -0,0 +1,331 @@
+=============================================================================
+
+Qiskit Dynamics provides model transformations and different numerical methods
+for evaluating models that can be used to customize how a simulation is done,
```
Line 5 is a bit confusing, when I think it through it makes sense. But if you could split it up a bit that might be nice

DanPuzzuoli replied:
Agreed, just expanded this into what I think is a better paragraph.

brosand commented on docs/userguide/how_to_configure_simulations.rst:
```diff
@@ -0,0 +1,331 @@
+with reasonable accuracy based on our specified tolerances.
+
+The discrepancy in solving times can be understood by examining the number of
+RHS evaluations when solving in each instance. The number of RHS evaluations for
```
Maybe define RHS here in case a new user isn't familiar with the terminology. Although maybe everyone who is using this package already has a strong understanding?

DanPuzzuoli replied:
Changed this to: `right-hand side (RHS)`, and added: "when solving the differential equation". I think this is a good addition, and _should_ be sufficient, all the key words are there. Also yeah, if a user is at the point of wanting to configure simulations they should know what these words mean at this point anyway.

brosand replied:
Agreed, great

brosand commented on docs/userguide/how_to_configure_simulations.rst:
```diff
@@ -0,0 +1,331 @@
+Reconstruct the model pieces at a much larger dimension, to observe the
+benefits of using sparse arrays. Furthermore, set up the initial state to
+be a single column vector, to
+further highlights the benefits of the sparse representation.
```
```suggestion
further highlight the benefits of the sparse representation.
```

brosand reviewed (approved):
A couple small edits, @chriseclectic may want to take a closer look as well, I would be interested to see other notes to see what I should look at in the future.

DanPuzzuoli commented:
@brosand I tried to address your comments by modifying the sections you suggested. 

I also needed to make a couple of other fixes:
- Switch to using the new `Array.set_default_backend` over the now-deprecated `dispatch.set_default_backend` in the userguide entry
- This change also still needed to be made in the `GeneratorModel` class, which was raising a further deprecation warning when setting the evaluation mode to `'sparse'`, which I've fixed now.
- Related to the previous point, I've added the word "Array" when referring to the default backend in descriptions of jax sparse evaluation mode.

brosand reviewed (approved):
LGTM
